keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/kep.yaml

```diff
@@ -7,8 +7,8 @@ reviewers: ['@thockin', '@danwinship', "@aojea"]
 approvers: ['@thockin']
 creation-date: "2023-02-03"
 status: implementable
-stage: alpha
-latest-milestone: "v1.28"
+stage: beta
```

Comment:
Few comments about the KEP itself:

> As such any vendor doing this would get a heads-up during Alpha.

People are not using Alpha.
Can we make any survey of providers to check who will be hit by it?
@thockin - thoughts?

> Implement change in Kube-proxy which will react to changes on the Node object and once the taint ToBeDeletedByClusterAutoscaler is placed on the Node object: start failing it's healthz state. 

I started wondering about this one. Namely if we shouldn't do that in two steps...
(a) introduce livez that [reacts to the existence of taint], but doesn't touch healthz
(b) graduate to stable and give people time and chance to use something stable
(c) [potentially] only then change healthz itself

@thockin - thoguhts?

> The metric: proxy_healthz_503_count mentioned in [Monitoring requirements]

I don't see that metric in the codebase - was this added?


> Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

has this happened? Can you update the KEP with the test and findings?


Please fill in the remaining questions in Throubleshooting section.

Reply:
> Can we make any survey of providers to check who will be hit by it?

Do we have a general procedure for that? I guess we could create a Google Form and send that out to dev@kubernetes.io, but that would still feel inconclusive.

 > I don't see that metric in the codebase - was this added?

Yes, here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/metrics/metrics.go#L189-L209

The name changed during the implementation however, because we decided to have the HTTP status code be used as a label. I will update the KEP with this

 > has this happened?

No

> Please fill in the remaining questions in Throubleshooting section.

OK, will update that now

Reply:
> People are not using Alpha.
> Can we make any survey of providers to check who will be hit by it?

I went back and re-read the whole KEP and scanned the code in question.  I agree that signal during alpha is ~0.  That said, I can't reason any way this explodes (other than logic bugs, obviously) - but I also know that this is super subtle stuff.

What would we do differently with more information, and as much I know the subtlety, I doubt anyone would look at this plan and see the explosion-path.  If this explodes, it's going to be something nobody can see until it's lit up.  :(

I think the mitigations in place are as low-risk as we can get with this, short of flipping the polarity entirely, which will almost guarantee low adoption.

>> Implement change in Kube-proxy which will react to changes on the Node object and once the taint ToBeDeletedByClusterAutoscaler is placed on the Node object: start failing it's healthz state.
>
> I started wondering about this one. Namely if we shouldn't do that in two steps...
> (a) introduce livez that [reacts to the existence of taint], but doesn't touch healthz
> (b) graduate to stable and give people time and chance to use something stable
> (c) [potentially] only then change healthz itself

This is the "flip the polarity" path, but it would not be `livez`.  More like "add `/healthz-v2`; get people to use it; deprecate old /healthz".  That is the safer path, for sure.  If we go that way, I don't think we will ever actually get rid of or fix the old healthz, and all of the existing clusters in the world will get no benefits.  Providers will be reluctant to adopt the new path because then they have to support the old AND the new, and who wants to do that?  

The lack of in-place updates sort of kills the value of this KEP, IMO, and the risk is not proportional to the reaction.  That said, this is not such a high-urgency thing that we should take on dramatic risk.  If you think I am wrong on the risk level, then we can try to go through a new URL endpoint and see if we can drive adoption.  That means figuring ouot, for example, if cloud-providers support updates to healthcheck configs.

Reply:
Thanks Tim - I guess I convinced by your explanation above. Let's just update the troubleshooting section as I requested and let's proceed with the plan